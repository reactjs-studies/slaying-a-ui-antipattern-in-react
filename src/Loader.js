// eslint-disable-next-line no-unused-vars
import React, { Component } from 'react';
import daggy from 'daggy';

const RemoteData = daggy.taggedSum('RemoteData', {
 NotAsked:[],
 Loading: [],
 Failure: ['error'],
 Success: ['data'],
});

class Loader extends Component {
 state = { items: RemoteData.NotAsked };

 loadData = () => {
    const { fetch } = this.props;
    this.setState({items: RemoteData.Loading});

    const getItems = async () => {
        let response;
        try {
            response = await fetch();
            this.setState({items:RemoteData.Success(response)})
        }catch (err){
            this.setState({items:RemoteData.Failure(err)});
        }
     }

     getItems();
 };

 getView = (items = {}) => {
     const {
           renderNotAsked,
           renderLoading,
           renderError,
           renderSuccess
         } = this.props;
     return items.cata({
           NotAsked: () => renderNotAsked(),
           Loading: () => renderLoading(),
           Failure: error => renderError(error),
           Success: data => renderSuccess(data)
         });
   };

 render(){ 
    const { items } = this.state;
    const { children } = this.props;
    const View = this.getView(items);
    return children(View, this.loadData);
  }
}

export default Loader;
