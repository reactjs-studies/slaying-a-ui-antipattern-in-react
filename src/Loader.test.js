import React from 'react';
import ReactDOM from 'react-dom';
import daggy from 'daggy';
import { shallow } from 'enzyme';

import  Loader  from './Loader';
import renderer from 'react-test-renderer';

const Enzyme = require('enzyme');
const Adapter = require('enzyme-adapter-react-16');
Enzyme.configure({ adapter: new Adapter() });

const $ = renderer.create;

const RemoteData = daggy.taggedSum('RemoteData', {
 NotAsked:[],
 Loading: [],
 Failure: ['error'],
 Success: ['data'],
});

const fetch = () => 
   new Promise ((res, rej) => {
       res([
            { id: 1, name: 'foo' },
            { id: 2, name: 'bar' },
            { id: 3, name: 'baz' }
       ])
   });

const props = {
 fetch,
 renderNotAsked: () => (<div>Not Loaded Data</div>),
 renderLoading:() =>  (<div>...loading</div>),
 renderError: (err) => (<div>{ err }</div>),
 renderSuccess:(data = []) => (<div>{ data.map(item => (<p key={item.id} >{item.name}</p>)) }</div>),
};

describe('Loader', ()=> {
    it('render not asked', () => {
        const tree = $(
            <Loader {...props}>
            {
                (view, LoadData) => (
                   <div>  
                    {view}
                    <button>Load data</button>
                   </div>
                )
            }
            </Loader>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('render loading data', () => {
        const wrapper = shallow(
            <Loader {...props}>
            {
                (view, LoadData) => (
                   <div>  
                    {view}
                    <button>Load data</button>
                   </div>
                )
            }
            </Loader>
        );

        wrapper.setState({items: RemoteData.Loading});
        expect(wrapper).toMatchSnapshot();
    });
    
    it('render data loaded', () => {
        const wrapper = shallow(
            <Loader {...props}>
            {
                (view, LoadData) => (
                   <div>  
                    {view}
                    <button>Load data</button>
                   </div>
                )
            }
            </Loader>
        );

        wrapper.setState({items: RemoteData.Success([{id:2, name:''}])});
        expect(wrapper).toMatchSnapshot();
    });

    it('data loading fail', () => {
        const wrapper = shallow(
            <Loader {...props}>
            {
                (view, LoadData) => (
                   <div>  
                    {view}
                    <button>Load data</button>
                   </div>
                )
            }
            </Loader>
        );

        wrapper.setState({items: RemoteData.Failure('error')});
        expect(wrapper).toMatchSnapshot();
    });
});

