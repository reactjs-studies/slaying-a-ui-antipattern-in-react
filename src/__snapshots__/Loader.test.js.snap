// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader data loading fail 1`] = `
<div>
  <div>
    error
  </div>
  <button>
    Load data
  </button>
</div>
`;

exports[`Loader render data loaded 1`] = `
<div>
  <div>
    <p
      key="2"
    />
  </div>
  <button>
    Load data
  </button>
</div>
`;

exports[`Loader render loading data 1`] = `
<div>
  <div>
    ...loading
  </div>
  <button>
    Load data
  </button>
</div>
`;

exports[`Loader render not asked 1`] = `
<div>
  <div>
    Not Loaded Data
  </div>
  <button>
    Load data
  </button>
</div>
`;
